src/mender-update/daemon/states.cpp

```diff
@@ -103,15 +103,49 @@ void IdleState::OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) {
 	log::Debug("Entering Idle state");
 }
 
+SubmitInventoryState::SubmitInventoryState(
+	events::EventLoop &event_loop, int retry_interval_seconds, int retry_count) :
+	retry_(Retry {
+		http::ExponentialBackoff(chrono::seconds(retry_interval_seconds), retry_count),
+		event_loop}) {
+}
+
+void SubmitInventoryState::HandlePollingError(Context &ctx, sm::EventPoster<StateEvent> &poster) {
+	auto exp_interval = retry_->backoff.NextInterval();
+	if (!exp_interval) {
+		log::Debug(
+			"Not retrying with backoff, retrying InventoryPollIntervalSeconds: "
+			+ exp_interval.error().String());
+		return;
+	}
+	log::Info(
+		"Retrying deployment polling in "
+		+ to_string(chrono::duration_cast<chrono::seconds>(*exp_interval).count()) + " seconds");
+
+	retry_->wait_timer.Cancel();
+	retry_->wait_timer.AsyncWait(*exp_interval, [&poster](error::Error err) {
+		if (err != error::NoError) {
+			if (err.code != make_error_condition(errc::operation_canceled)) {
+				log::Error("Retry poll timer caused error: " + err.String());
+			}
+		} else {
+			poster.PostEvent(StateEvent::DeploymentPollingTriggered);
+		}
+	});
+}
+
 void SubmitInventoryState::DoSubmitInventory(Context &ctx, sm::EventPoster<StateEvent> &poster) {
 	log::Debug("Submitting inventory");
 
-	auto handler = [&ctx, &poster](error::Error err) {
+	auto handler = [this, &ctx, &poster](error::Error err) {
 		if (err != error::NoError) {
+			// Replace the inventory poll timer with a backoff
+			HandlePollingError(ctx, poster);
```

Comment:
Here (and below) I think it is best if we first print the error message (next line) and then we handle it (where it will log that it will be retried in a short while)